src/graph/unit_graph.cc

```diff
@@ -881,9 +987,17 @@ HeteroGraphPtr UnitGraph::AsNumBits(HeteroGraphPtr g, uint8_t bits) {
     //   be fixed later.
     auto bg = std::dynamic_pointer_cast<UnitGraph>(g);
     CHECK_NOTNULL(bg);
-    CSRPtr new_incsr = CSRPtr(new CSR(bg->GetInCSR()->AsNumBits(bits)));
-    CSRPtr new_outcsr = CSRPtr(new CSR(bg->GetOutCSR()->AsNumBits(bits)));
-    return HeteroGraphPtr(new UnitGraph(g->meta_graph(), new_incsr, new_outcsr, nullptr));
+
+    if (!bg->prefer_coo_) {
+      CSRPtr new_incsr = CSRPtr(new CSR(bg->GetInCSR()->AsNumBits(bits)));
+      CSRPtr new_outcsr = CSRPtr(new CSR(bg->GetOutCSR()->AsNumBits(bits)));
+      return HeteroGraphPtr(
+          new UnitGraph(g->meta_graph(), new_incsr, new_outcsr, nullptr, false));
+    } else {
+      COOPtr new_coo = COOPtr(new COO(bg->GetCOO()->AsNumBits(bits)));
+      return HeteroGraphPtr(
+          new UnitGraph(g->meta_graph(), nullptr, nullptr, new_coo, true));
+    }
```

Comment:
The question is for any function `GG = foo(G)`, if `G` is a hypersparse graph (preferring COO), should `GG` also prefer `COO`? I think maybe a better solution is to lower the `detect_prefer_coo` function in python to C++ so whenever a graph is created, it will try to store it in a more efficient format.

Reply:
It is lowered in C++ now.